pkg/copyutil/copyutil.go

```diff
@@ -0,0 +1,96 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package copyutil
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+// copied from https://github.com/kubernetes/kubectl/blob/v0.23.4/pkg/cmd/cp/filespec.go#L126-L134 (v0.23.4)
+func StripTrailingSlash(file string) string {
+	if len(file) == 0 {
+		return file
+	}
+	if file != "/" && strings.HasSuffix(string(file[len(file)-1]), "/") {
+		return file[:len(file)-1]
+	}
+	return file
+}
+
+// SplitPathDirEntry splits the given path between its directory name and its
+// basename. We assume that remote files are both linux and windows containers.
+// So we first replace each slash ('/') character in path with a separator character
+// then we clean the path.
+// A trailing "." is preserved if the original path specified the current directory.
+func SplitPathDirEntry(path string) (dir, base string) {
```

Comment:
https://github.com/search?q=SplitPathDirEntry&type=code

When you copied codes from somewhere else, please confirm the origin and the license.